In [39]:
import numpy as np
import pandas as pd
from pathlib import Path
import matplotlib.pyplot as plt
from pathlib import Path
from collections import Counter
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder
from sklearn.ensemble import RandomForestClassifier
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd

In [40]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

Cleaning up training dataset

In [41]:
train_df.head(2)

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0       57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1      141451  141451    21000.0    0.1308       478.68       MORTGAGE   

   annual_inc verification_status loan_status pymnt_plan  ...  pct_tl_nvr_dlq  \
0    223000.0        Not Verified    low_risk          n  ...           100.0   
1    123000.0     Source Verified    low_risk          n  ...            85.0   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              50.0                   0.0        0.0         577150.0   
1              33.3                   0.0        0.0         132750.0   

   total_bal_ex_mort  total_bc_limit total_il_high_credit_limit  \
0           122018.0         32000.0                   170200.0   
1            27896.0         15900.0                    35398.0   

   hardship_flag  debt_settlement_flag  
0              N                     N  
1              N                     N  

[2 rows x 86 columns]

In [42]:
train_df["loan_status"].value_counts()

low_risk     6090
high_risk    6090
Name: loan_status, dtype: int64

In [43]:
#defining target variable
y_train1=train_df["loan_status"] #defining label
y_train1

0         low_risk
1         low_risk
2         low_risk
3         low_risk
4         low_risk
           ...    
12175    high_risk
12176    high_risk
12177    high_risk
12178    high_risk
12179    high_risk
Name: loan_status, Length: 12180, dtype: object

In [44]:
# Convert categorical data to numeric and separate target feature for training data
y_train = LabelEncoder().fit_transform(y_train1)
y_train

# we can see lower risk is 1, high risk is 0 in our data

array([1, 1, 1, ..., 0, 0, 0])

In [45]:
#drop some cloumns 
clean_traindf=train_df.drop(['loan_status','Unnamed: 0','index'], axis=1) 
clean_traindf.head(2)

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    13375.0    0.1797       483.34       MORTGAGE    223000.0   
1    21000.0    0.1308       478.68       MORTGAGE    123000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Not Verified          n  29.99          0.0             0.0  ...   
1     Source Verified          n  11.26          2.0             0.0  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0           100.0              50.0                   0.0        0.0   
1            85.0              33.3                   0.0        0.0   

  tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0        577150.0           122018.0         32000.0   
1        132750.0            27896.0         15900.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                    170200.0              N                     N  
1                     35398.0              N                     N  

[2 rows x 83 columns]

In [46]:
# Convert categorical data to numeric and separate target feature for training data
X_train=pd.get_dummies(clean_traindf)
print(X_train)

       loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0        13375.0    0.1797       483.34    223000.0  29.99          0.0   
1        21000.0    0.1308       478.68    123000.0  11.26          2.0   
2        20000.0    0.1240       448.95    197000.0  11.28          0.0   
3         3000.0    0.1240       100.22     45000.0  18.08          0.0   
4        30000.0    0.1612      1056.49    133000.0  27.77          0.0   
...          ...       ...          ...         ...    ...          ...   
12175    19975.0    0.2565       801.09     28000.0  28.42          0.0   
12176    15000.0    0.1774       540.34     50000.0  23.43          4.0   
12177     3600.0    0.1862       131.28     60000.0  28.80          0.0   
12178    15000.0    0.0881       475.68     62000.0  11.44          0.0   
12179    15000.0    0.1774       540.34     54080.0  21.66          1.0   

       inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0                 0.0      15.0      0.

In [47]:
clm=X_train.columns
clm

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

In [48]:
print(X_train.columns)

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

In [49]:
X_train.head(2)

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    13375.0    0.1797       483.34    223000.0  29.99          0.0   
1    21000.0    0.1308       478.68    123000.0  11.26          2.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             0.0      15.0      0.0    39728.0  ...   
1             0.0      16.0      0.0     9585.0  ...   

   verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                             0             1                      0   
1                             0             1                      0   

   initial_list_status_w  application_type_Individual  \
0                      1                            1   
1                      1                            1   

   application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                           0                1                0   
1                           0                1                0   

   debt_settlement_flag_N  debt_settlement_flag_Y  
0                       1                       0  
1                       1                       0  

[2 rows x 92 columns]

In [50]:

print(X_train.shape, y_train.shape)

(12180, 92) (12180,)


Cleaning up test dataset

In [51]:
test_df.head(2)

Unnamed: 0  index  loan_amnt  int_rate  installment home_ownership  \
0       67991  67991    40000.0    0.0819        814.7       MORTGAGE   
1       25429  25429     6000.0    0.1524        208.7           RENT   

   annual_inc verification_status loan_status pymnt_plan  ...  pct_tl_nvr_dlq  \
0    140000.0        Not Verified    low_risk          n  ...            97.7   
1     55000.0        Not Verified    low_risk          n  ...            66.7   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0               0.0                   0.0        0.0         527975.0   
1               0.0                   0.0        0.0          34628.0   

   total_bal_ex_mort  total_bc_limit total_il_high_credit_limit  \
0            70914.0         74600.0                    99475.0   
1            23460.0          5900.0                    23628.0   

   hardship_flag  debt_settlement_flag  
0              N                     N  
1              N                     N  

[2 rows x 86 columns]

In [52]:
#defining y variable
y_test1=test_df["loan_status"] #defining label
y_test1

0        low_risk
1        low_risk
2        low_risk
3        low_risk
4        low_risk
          ...    
4697    high_risk
4698    high_risk
4699    high_risk
4700    high_risk
4701    high_risk
Name: loan_status, Length: 4702, dtype: object

In [53]:
# Convert categorical data to numeric and separate target feature for testing data
y_test = LabelEncoder().fit_transform(y_test1)
y_test

array([1, 1, 1, ..., 0, 0, 0])

In [54]:
#drop some cloumns 
clean_testdf=test_df.drop(['loan_status','Unnamed: 0','index'], axis=1) 
clean_testdf.head(2)

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    40000.0    0.0819        814.7       MORTGAGE    140000.0   
1     6000.0    0.1524        208.7           RENT     55000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Not Verified          n  19.75          0.0             1.0  ...   
1        Not Verified          n  11.52          2.0             0.0  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            97.7               0.0                   0.0        0.0   
1            66.7               0.0                   0.0        0.0   

  tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0        527975.0            70914.0         74600.0   
1         34628.0            23460.0          5900.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     99475.0              N                     N  
1                     23628.0              N                     N  

[2 rows x 83 columns]

In [55]:
clean_testdf.dropna(inplace=True)

In [56]:
#test dataset, converting values
X_test=pd.get_dummies(clean_testdf)
print(X_test.columns)

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

In [57]:
X_test.head(2)

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    40000.0    0.0819        814.7    140000.0  19.75          0.0   
1     6000.0    0.1524        208.7     55000.0  11.52          2.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             1.0      18.0      0.0     9471.0  ...   
1             0.0       8.0      0.0     1280.0  ...   

   verification_status_Source Verified  verification_status_Verified  \
0                                    0                             0   
1                                    0                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  
0                0                       1  
1                0                       1  

[2 rows x 91 columns]

In [58]:

print(X_test.shape, y_test.shape)

(4702, 91) (4702,)


In [59]:
#find missing columns
column=set(X_train.columns)-set(X_test.columns)
column

{'debt_settlement_flag_Y'}

In [60]:
X_test.insert(91,'debt_settlement_flag_Y',0)

In [61]:
X_train.debt_settlement_flag_Y.value_counts()

0    12175
1        5
Name: debt_settlement_flag_Y, dtype: int64

In [62]:
print(X_test.shape,y_test.shape)

(4702, 92) (4702,)


Drop missing column from train data wheich as added to testing data in previous code

In [63]:
X_train.drop(['debt_settlement_flag_Y'], axis=1, inplace=True ) 
X_train.head(2)

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    13375.0    0.1797       483.34    223000.0  29.99          0.0   
1    21000.0    0.1308       478.68    123000.0  11.26          2.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             0.0      15.0      0.0    39728.0  ...   
1             0.0      16.0      0.0     9585.0  ...   

   verification_status_Source Verified  verification_status_Verified  \
0                                    0                             0   
1                                    1                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  
0                0                       1  
1                0                       1  

[2 rows x 91 columns]

In [64]:
X_test.drop(['debt_settlement_flag_Y'], axis=1, inplace=True ) 
X_test.head()

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    40000.0    0.0819       814.70    140000.0  19.75          0.0   
1     6000.0    0.1524       208.70     55000.0  11.52          2.0   
2     3600.0    0.1695       128.27     42000.0   6.74          0.0   
3    20000.0    0.1524       478.33    100000.0  12.13          0.0   
4     3600.0    0.1240       120.27     50000.0  16.08          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             1.0      18.0      0.0     9471.0  ...   
1             0.0       8.0      0.0     1280.0  ...   
2             0.0       6.0      0.0     4757.0  ...   
3             2.0       7.0      0.0    12731.0  ...   
4             3.0       6.0      0.0    10413.0  ...   

   verification_status_Source Verified  verification_status_Verified  \
0                                    0                             0   
1                                    0                             0   
2                                    0                             0   
3                                    0                             0   
4                                    0                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   
2                            1                           0                1   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  
0                0                       1  
1                0                       1  
2                0                       1  
3                0                       1  
4                0                       1  

[5 rows x 91 columns]

Before the analysis I think random forrest will be better predictor.

Logistic regression unscaled data

In [81]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression
classifier=LogisticRegression(max_iter=2000,solver = 'lbfgs')
classifier

LogisticRegression(max_iter=2000)

In [82]:
classifier.fit(X_train, y_train)

C:\Users\MDavi23\Conda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=2000)

In [83]:
print(f"Logistic Regression unscaled data")
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Logistic Regression unscaled data
Training Data Score: 0.6980295566502464
Testing Data Score: 0.5776265418970651


Random Forrest Classifier

In [71]:
# Train a Random Forest Classifier model and print the model score
rfc= RandomForestClassifier(random_state=1, n_estimators=500)
rfc

RandomForestClassifier(n_estimators=500, random_state=1)

In [72]:
rfc.fit(X_train, y_train)

RandomForestClassifier(n_estimators=500, random_state=1)

In [73]:
print(f"Random Forest Classifier unscaled data")
print(f'Training Score: {rfc.score(X_train, y_train)}')
print(f'Testing Score: {rfc.score(X_test, y_test)}')

Random Forest Classifier unscaled data
Training Score: 1.0
Testing Score: 0.647809442790302


In [74]:
# Scale the data
scaler = StandardScaler().fit(X_train)
X_train_scaleddata = scaler.transform(X_train)
X_test_scaleddata= scaler.transform(X_test)

Logistic Regression with Scaled Data

In [76]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression
classifier=LogisticRegression(max_iter=2000,solver = 'lbfgs')
classifier



classifier.fit(X_train_scaleddata, y_train)

LogisticRegression(max_iter=2000)

In [77]:
print(f"Logistic Regression scaled data")
print(f"Training Data Score: {classifier.score(X_train_scaleddata, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test_scaleddata, y_test)}")

Logistic Regression scaled data
Training Data Score: 0.7079638752052545
Testing Data Score: 0.7679710761378137


Random Forrest with Scaled Data

In [79]:
# Train a Random Forest Classifier model on the scaled data and print the model score

rfc = RandomForestClassifier(n_estimators=100, random_state=1)
rfc.fit(X_train_scaleddata, y_train)

RandomForestClassifier(random_state=1)

In [80]:
print(f"Random Forest Classifier scaled data")
print(f'Training Score: {rfc.score(X_train_scaleddata, y_train)}')
print(f'Testing Score: {rfc.score(X_test_scaleddata, y_test)}')

Random Forest Classifier scaled data
Training Score: 1.0
Testing Score: 0.6401531263292216


Analysis for Logistic Regression and Random Forest Classifier Model for unscaled data

Logistic regression is lowest  for unscaled data for both training and testing data.
Logistic Regression unscaled data
Training Data Score: 0.6980295566502464
Testing Data Score: 0.5776265418970651


Random Forest Classifier model is in max for training data, low for testing data.
Random Forest Classifier unscaled data
Training Score: 1.0
Testing Score: 0.647809442790302




Analysis for Logistic Regression and Random Forest Classifier Model for scaled data

Logistic regression with scaled data is slightly higher than logistic regression with unscaled data for both training and testing data.

Training Data Score: 0.7079638752052545 Testing Data Score: 0.7679710761378137

Random Forest Classifier model is in max for training data as unscaled data, still pretty low for testing data as unscaled data.

Training Score: 1.0 Testing Score: 0.6401531263292216


